// @mayahq/module-sdk/src/storage/runtimeStorage.js
const path = require('path')
const fs = require('fs')
const update = require('immutability-helper')

const Storage = require('./storage')
const { execSetQuery, execGetQuery } = require('./query')

class RuntimeStorage extends Storage {
    constructor({ root, runtimeId }) {
        super({ root, type: 'RUNTIME' })
        this.runtimeId = runtimeId

        this.dbpath = path.join(root, 'persistentStore', runtimeId, 'global.json')
        this.lockpath = path.join(root, 'persistentStore', runtimeId, 'global.lock')

        const runtimeData = Storage._get(this.dbpath, {})
        this.data = runtimeData
    }

    get(query) {
        return Storage._executeSyncOperationOnPath(this.lockpath, () => {
            const data = require(this.dbpath)
            const result = execGetQuery(data, query)
            return result
        })
    }

    set(query) {
        return Storage._executeSyncOperationOnPath(this.lockpath, () => {
            const data = require(this.dbpath)
            const result = execSetQuery(data, query)
            fs.writeFileSync(this.dbpath, JSON.stringify(result))
            return result
        })
    }

    update(updates) {
        return Storage._executeSyncOperationOnPath(this.lockpath, () => {
            const data = require(this.dbpath)
            const result = update(data, updates)
            fs.writeFileSync(this.dbpath, JSON.stringify(result))
            return result
        })
    }
}

module.exports = RuntimeStorage
