// @mayahq/maya-db/lib/src/try.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = __importDefault(require("path"));
var index_1 = require("./index");
var dbroot = path_1.default.resolve(process.env.HOME, 'maya/testdb1');
var db = (0, index_1.localDb)({
    encryptionKey: 'eda344e1ab8b9e122aab3350eec33e95802c7fe68aac8ad85c5c64d97e45ef1a',
    root: dbroot
});
var targetJson = {
    name: 'Dushyant',
    education: {
        college: 'BITS Pilani',
        graduation: 2022,
        degree: {
            type: 'single',
            branch: 'CS'
        }
    },
    address: 'C-154 HKM Nagar'
};
var tree = {
    tokens: [
        {
            safe: [
                {
                    spotify: 'BLOCK',
                },
                {
                    google: 'ENCRYPTED_BLOCK',
                },
                {
                    testcol: []
                }
            ]
        },
        {
            unsafe: [
                {
                    spotify: 'ENCRYPTED_BLOCK',
                },
                {
                    testcol: []
                }
            ]
        }
    ],
    global: 'BLOCK'
};
// console.log(db.collection('global').getAllBlocks())
db.ensureHierarchy(tree);
// const block = db.createNewCollection('global').createNewBlock('spotify', {encrypted: false})
// const block = db.block('global/spotify')
// const block = db.collection('global').block('spotify')
// block.set(targetJson, {
//     overwrite: true
// })
// const result = block.get({
//     name: null,
//     lastName: 'Jain',
//     education: {
//         college: null,
//         degreeType: 'single'
//     }
// }).then((result) => {
//     console.log(result)
// })
// {
//     name: 'Dushyant',
//     lastName: 'Jain',
//     education: {
//         college: 'BITS Pilani',
//         degreeType: 'single'
//     }
// }
var tree2 = {
    tokens: [
        {
            oauth: [
                {
                    v2: [
                        { spotify: 'ENCRYPTED_BLOCK' },
                        { google: 'ENCRYPTED_BLOCK' }
                    ]
                },
                {
                    v1: [
                        { hubspot: 'ENCRYPTED_BLOCK' }
                    ]
                }
            ]
        }
    ],
    global: [
        {
            profiles: [
                { module1: 'BLOCK' },
                { module2: 'BLOCK' }
            ]
        }
    ]
};
