// @mayahq/maya-db/lib/src/io/mongo/blockSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MayaDbBlockSchema = void 0;
var mongoose_1 = __importDefault(require("mongoose"));
var path_1 = __importDefault(require("path"));
exports.MayaDbBlockSchema = new mongoose_1.default.Schema({
    path: {
        type: String,
        unique: true,
        required: true,
        immutable: true,
        index: true
    },
    parentPath: {
        type: String,
        index: true
    },
    lockAcquiredBy: {
        type: String,
        default: ''
    },
    lockExpiresAt: {
        type: Number,
        default: -1
    },
    data: {
        type: String,
        default: '{}',
        validate: {
            validator: function (v) {
                try {
                    JSON.stringify(v);
                    return true;
                }
                catch (e) {
                    return false;
                }
            },
            message: function () { return 'Data must be specified in JSON format'; }
        }
    }
});
exports.MayaDbBlockSchema.pre('save', function (next) {
    var parent = path_1.default.dirname(this.path);
    if (parent === '.') {
        parent = '/';
    }
    this.parentPath = parent;
    next();
});
var MayaDbBlock = mongoose_1.default.model('MayaDbBlock', exports.MayaDbBlockSchema);
exports.default = MayaDbBlock;
